### PR TITLE
Fix the auth helpers to only require fields they need from `AuthIdentity`

### PR DESCRIPTION
**Pull request by infomiho**
The auth helpers like `getEmail` and `getUsername` didn't work unless you provided the full user object with full `AuthIdentity` objects in the `identities` array.

Given that we[ suggest to users to include only the `providerName` and `providerUserId`](https://github.com/wasp-lang/wasp/blob/4bb0f9e50d6028681e3d3c6e9ffd599edc455d72/web/docs/auth/entities/entities.md?plain=1#L212) when manually fetching the user + auth + identities, we should make sure our auth helpers work in that scenario. 

This PR:
- makes it so that they only need to pass in the `providerName` and `providerUserId` when using the helpers.
- includes a headless test to ensure this helper works in the future.

### CODE REVIEW

File: waspc/data/Generator/templates/sdk/wasp/server/auth/user.ts [hunk 1]
```diff
@@ -63,6 +63,23 @@ export type AuthEntityWithIdentities = {= authEntityName =} & {
   {= identitiesFieldOnAuthEntityName =}: {= authIdentityEntityName =}[]
 }
 
+// PRIVATE API
```
Comment (infomiho) [inline waspc/data/Generator/templates/sdk/wasp/server/auth/user.ts:66]: This is basically a copy of the types we used so far, but it reduces the needed fields from _all the AuthIdentity fields_ to only the ones we really need.

File: waspc/src/Wasp/Generator/SdkGenerator/AuthG.hs [hunk 2]
```diff
@@ -30,7 +30,7 @@ genAuth spec =
     Just auth ->
       -- shared stuff
       sequence
-        [ genUserTs
+        [ genFileCopy [relfile|auth/user.ts|]
```
Comment (infomiho) [inline waspc/src/Wasp/Generator/SdkGenerator/AuthG.hs:33]: No need for templating anymore for this file, I've used a copy file draft instead.

File: waspc/headless-test/examples/todoApp/src/server/queries.ts [hunk 3]
```diff
@@ -1,29 +1,54 @@
...
+              },
+            },
+          },
+        },
+      },
+    },
+  })
   return tasks
-}
+}) satisfies GetTasks<void>
```
Comment (sodic) [inline waspc/headless-test/examples/todoApp/src/server/queries.ts:39]: How come the frontend type tests aren't complaining about this?

**Comment by infomiho** [comment]
Good point! The tricky part with the `UserEntityWithAuth` name is that it is correct in a sense that it represents a tree of full Prisma entities. The second type is derived from it by making it _as minimal as possible_ to make the function as flexible as possible. 

The problem is that the user will see this wacky (albeit correct) name. 

Some alternatives:
- internal: `UserEntityWithFullAuth`, public: `UserEnittyWithAuth`
- internal `AuthEntityTree`, public: `UserEnittyWithAuth`
- internal: `UserEntityWithAuthFields`, public: `UserWithRequiredAuthFields`

I maybe like the first option the best since it's just a `Full` vs nothing which gives us the relationship you suggested.

File: waspc/data/Generator/templates/sdk/wasp/auth/user.ts [hunk 4]
```diff
@@ -49,7 +51,7 @@ function makeAuthUser(data: AuthUserData): AuthUser {
   };
 }
 
-function findUserIdentity(user: UserEntityWithAuth, providerName: ProviderName): {= authIdentityEntityName =} | null {
+function findUserIdentity(user: UserEntityWithAuth, providerName: ProviderName): UserEntityWithAuth['auth']['identities'][number] | null {
```
Comment (sodic) [inline waspc/data/Generator/templates/sdk/wasp/auth/user.ts:54]: Nice!

File: waspc/headless-test/examples/todoApp/src/client/Todo.tsx [hunk 5]
```diff
@@ -130,6 +132,8 @@ const TaskView = ({ task }: { task: Task }) => {
     }
   }
 
+  const email = getEmail(task.user)
```
[the remark this answers is not in the document]
Comment (sodic) [reply waspc/headless-test/examples/todoApp/src/client/Todo.tsx:135]: What do you mean here?

**Review by sodic — APPROVED** [review]
Exactly what I had in mind, nice work!

File: waspc/headless-test/examples/todoApp/src/client/Todo.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (infomiho) [reply waspc/headless-test/examples/todoApp/src/client/Todo.tsx:135]: Using the `getEmail` helper in headless tests app to test it in headless tests :)

**Pull request merged**